src/dex/gmx/fast-price-feed.ts

```diff
@@ -0,0 +1,228 @@
+import { Interface } from '@ethersproject/abi';
+import { DeepReadonly } from 'ts-essentials';
+import { PartialEventSubscriber } from '../../composed-event-subscriber';
+import {
+  Address,
+  MultiCallInput,
+  MultiCallOutput,
+  Logger,
+  Log,
+  BlockHeader,
+} from '../../types';
+import { PoolState, FastPriceFeedConfig, FastPriceFeedState } from './types';
+import FastPriceFeedAbi from '../../abi/gmx/fast-price-feed.json';
+import FastPriceEventsAbi from '../../abi/gmx/fast-price-events.json';
+import { Lens } from '../../lens';
+
+export class FastPriceFeed<State> extends PartialEventSubscriber<
+  State,
+  FastPriceFeedState
+> {
+  static readonly interface = new Interface(FastPriceFeedAbi);
+  static readonly fastPriceEventsInterface = new Interface(FastPriceEventsAbi);
+
+  BASIS_POINTS_DIVISOR = 10000n;
+  protected priceDuration: number;
+  protected maxDeviationBasisPoints: bigint;
+  protected favorFastPrice: boolean;
+  protected volBasisPoints: bigint;
+
+  constructor(
+    private fastPriceFeedAddress: Address,
+    private fastPriceEventsAddress: Address,
+    private tokenAddresses: Address[],
+    config: FastPriceFeedConfig,
+    lens: Lens<DeepReadonly<State>, DeepReadonly<FastPriceFeedState>>,
+    logger: Logger,
+  ) {
+    super([fastPriceEventsAddress], lens, logger);
+    this.priceDuration = config.priceDuration;
+    this.maxDeviationBasisPoints = config.maxDeviationBasisPoints;
+    this.favorFastPrice = config.favorFastPrice;
+    this.volBasisPoints = config.volBasisPoints;
+  }
+
+  public processLog(
+    state: DeepReadonly<FastPriceFeedState>,
+    log: Readonly<Log>,
+    blockHeader: Readonly<BlockHeader>,
+  ): DeepReadonly<FastPriceFeedState> | null {
+    try {
+      const parsed = FastPriceFeed.fastPriceEventsInterface.parseLog(log);
+      switch (parsed.name) {
+        case 'PriceUpdate': {
+          const _state: FastPriceFeedState = state;
```

Comment:
You need to make a deep copy if you're going to modify the state, otherwise you are invalidating the state history!  (It's DeepReadonly for a reason...)